### ISSUE TITLE
smart negative indices

### ISSUE THREAD

**Issue by valq7711** [kovidgoyal/rapydscript-ng#80]
converting  all expressions like  ```obj[a][b][c]``` to the following one seems too costly
```javascript
(ρσ_expr_temp = (ρσ_expr_temp = obj[(typeof a === "number" && a < 0) ? obj.length + a : a])[(typeof b === "number" && b < 0) ? ρσ_expr_temp.length + b : b])[(typeof c === "number" && c < 0) ? ρσ_expr_temp.length + c : c];
};
```

Kovid, could you provide compile option like 'smart-negative-indices', please. So, when it is off:
 `obj[a][-1][c]` => `obj[a][ obj[a].length-1 ][c]` 
 i.e.  convert only explicit numbers

**Comment by kovidgoyal** [comment]
Do you have some actualy numbers showing it is significantly slower? IIRC I found the overhead for this was pretty low when I implemented it. Unfortunately, I did not keep the benchmarking results around.

**Comment by kovidgoyal** [comment]
I agree, it's ugly. But the python compatibility use-case trumps prettiness considerations. Any language design is about tradeoffs, this is one of them. And note that most of the ugliness comes from ρσ_expr_temp which will need to be used even for plain negative numbers. This is because the [] operator can have side-effects in JS (and python), so it must not be evaluated more times than in the original expression.
